### DISCUSSION TITLE
Native support for CockroachDB

### DISCUSSION THREAD

**Discussion by ChuckMoe** [fief-dev/fief#73] [Q&A]
Good evening together,

I would like to use [CockroachDB](https://cockroachlabs.com/) in my project. It is very easy to use it with SQLAlchemy as you can see [here](https://www.cockroachlabs.com/docs/v22.1/build-a-python-app-with-cockroachdb-sqlalchemy). All it needs is the installation of a python package and the dialect 'cockroachdb'.
My question now is, how you I achieve this elegantly? The only way that comes to mind currently is to fork the project, add the dependency and add COCKROACHDB as an allowed DATABASE_TYPE.

Cheers
Moe

**Comment by frankie567** [comment]
Hi @ChuckMoe 👋 

I would love to add support to CockroachDB! Actually, I've tried some time ago but didn't push forward as I got some issues with it. To add support, several things would be involved:

* Add it as a new type in `DatabaseType`: https://github.com/fief-dev/fief/blob/a45fd7abb1faa25dd8c5e9d9e940344b5b2ad15f/backend/fief/db/types.py#L9-L12
* Define its SSL modes: https://github.com/fief-dev/fief/blob/a45fd7abb1faa25dd8c5e9d9e940344b5b2ad15f/backend/fief/db/types.py#L15-L37
    * Add corresponding tests: https://github.com/fief-dev/fief/blob/d950f240016a196d269518a968243a5e471571f2/backend/tests/test_db_types.py
* Define its drivers: https://github.com/fief-dev/fief/blob/a45fd7abb1faa25dd8c5e9d9e940344b5b2ad15f/backend/fief/db/types.py#L39-L49 — That's probably the tricky part, because I remember asyncpg had issues with Cockroach
* Add it in the text matrix: https://github.com/fief-dev/fief/blob/25494565ca468108ea7b48192fbd3a5d2e37613e/.github/workflows/build.yml#L46-L67
* Check that the tests are passing, in particular the migrations.

From my experience, this can be quite complex and frustrating. But I would be glad if someone could have a first try with this.

**Reply by ChuckMoe** [reply]
Hi @frankie567  :wave:

Setting up CockroachDB with the right certificates and a useraccount, at least automatically, is bothersome. There is a [docker image](https://hub.docker.com/r/cockroachdb/cockroach/#!) you can use but it isn't documented, so I don't know if or how I could set an initial user.



>     * Define its drivers: https://github.com/fief-dev/fief/blob/a45fd7abb1faa25dd8c5e9d9e940344b5b2ad15f/backend/fief/db/types.py#L39-L49
>        — That's probably the tricky part, because I remember asyncpg had issues with Cockroach
They also state that support for asyncpg is still [experimental](https://github.com/cockroachdb/sqlalchemy-cockroachdb), yeah.

I will try do dig a little deeper. Can't say exactly when though..

**Comment by frankie567** [comment]
Had a stab at this myself in [branch cockroachdb-support](https://github.com/fief-dev/fief/tree/cockroachdb-support) and quickly ran into issues with `asyncpg` (can't even create a workspace).

It looks like issue https://github.com/cockroachdb/cockroach/issues/71908 is the root cause.

Anyway, this is rather promising and I think it could be quickly implemented when `asyncpg` support will be more stable.

**Reply by ChuckMoe** [reply]
In theory it's straight forward but I also quickly noticed that it won't work in its current state. I'll watch the issue and await the fix.
Thank you once more :)

**Comment by frankie567** [comment]
Apparently, CockroachDB team fixed the issue with asyncpg: https://github.com/cockroachdb/cockroach/pull/95234

Who knows, maybe we'll be able to going forward with this when they release it!

**Reply by ChuckMoe** [reply]
I got a notification of this feature merge as well! I am really curious on the next release!
